### PR TITLE
CBL-449: Pending Document ID ObjC

### PR DESCRIPTION
**Pull request by jayahariv**
* add implementation for getting the Pending Document ID in ObjC
* add one unit test

### CODE REVIEW

File: Objective-C/CBLReplicator.mm [hunk 1]
```diff
@@ -454,6 +455,33 @@ - (void) removeChangeListenerWithToken: (id<CBLListenerToken>)token {
...
+            *error = [NSError errorWithDomain: CBLErrorDomain
+                                         code: CBLErrorUnsupported
+                                     userInfo: @{NSLocalizedDescriptionKey: kCBLErrorMessagePullOnlyPendingDocIDs}];
+        return [NSSet set];
+    }
+    
+    
+    if (!_repl)
+        return [NSSet set];
+
```
Comment (jayahariv) [inline Objective-C/CBLReplicator.mm]:
```suggestion
```

File: Objective-C/CBLReplicator.mm [hunk 2]
```diff
@@ -454,6 +455,32 @@ - (void) removeChangeListenerWithToken: (id<CBLListenerToken>)token {
...
 
+- (NSSet<NSString*>*) pendingDocumentIds: (NSError**)error {
+    if (_config.replicatorType > 1) {
+        if (error)
+            *error = [NSError errorWithDomain: CBLErrorDomain
+                                         code: CBLErrorUnsupported
+                                     userInfo: @{NSLocalizedDescriptionKey: kCBLErrorMessagePullOnlyPendingDocIDs}];
+        return [NSSet set];
+    }
+
```
Comment (jayahariv) [inline Objective-C/CBLReplicator.mm]:
```suggestion
```

File: Objective-C/Tests/ReplicatorTest+PendingDocIds.m [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    
+    doc = [self createDocument: @"4"];
+    [doc setString:@"Smokey" forKey:@"name"];
+    [self saveDocument: doc];
+    
+    doc = [self createDocument: @"5"];
+    [doc setString:@"Smokey" forKey:@"name"];
+    [self saveDocument: doc];
+    
+
```
Comment (jayahariv) [inline Objective-C/Tests/ReplicatorTest+PendingDocIds.m]:
```suggestion
```

File: Objective-C/CBLReplicator.mm [hunk 4]
```diff
@@ -454,6 +455,36 @@ - (void) removeChangeListenerWithToken: (id<CBLListenerToken>)token {
     }
 }
 
+- (NSSet<NSString*>*) pendingDocumentIds: (NSError**)error {
+    if (_config.replicatorType > 1) {
+        if (error)
```
Comment (bmeike) [inline Objective-C/CBLReplicator.mm]: Is this right?  If there is already an error, replace it?

Comment (jayahariv) [reply Objective-C/CBLReplicator.mm]: this is to make sure, the passed in argument to this function is valid, then only we will assign value to the pointed location.

Comment (bmeike) [reply Objective-C/CBLReplicator.mm]: All good, then!

File: Objective-C/CBLReplicator.mm [hunk 5]
```diff
@@ -454,6 +455,38 @@ - (void) removeChangeListenerWithToken: (id<CBLListenerToken>)token {
...
+            *error = [NSError errorWithDomain: CBLErrorDomain
+                                         code: CBLErrorUnsupported
+                                     userInfo: @{NSLocalizedDescriptionKey: kCBLErrorMessagePullOnlyPendingDocIDs}];
+        return [NSSet set];
+    }
+    
+    if (!_repl) {
+        CBLLogInfo(Sync, @"Trying to fetch pending documentIds without a c4replicator %@", _repl);
+        return [NSSet set];
```
Comment (pasin) [inline Objective-C/CBLReplicator.mm]: I understand that we should be able to return the pending docs even when before starting the replicator and when the replicator is stopped. We probably need to change to be able to create the replicator on demand (lazy) and make sure to also free them in destructor.

File: Objective-C/CBLReplicator.mm [hunk 6]
```diff
@@ -454,6 +455,38 @@ - (void) removeChangeListenerWithToken: (id<CBLListenerToken>)token {
...
+    
+    C4Error c4err = {};
+    C4SliceResult result = c4repl_getPendingDocIDs(_repl, &c4err);
+    if (c4err.code > 0) {
+        convertError(c4err, error);
+        CBLWarnError(Sync, @"Error while fetching pending documentIds: %d/%d", c4err.domain, c4err.code);
+        return [NSSet set];
+    }
+    
+    if (!result.size)
```
Comment (pasin) [inline Objective-C/CBLReplicator.mm]: I feel easier to understand when doing result.size > 0.

Comment (jayahariv) [reply Objective-C/CBLReplicator.mm]:
```suggestion
    if (result.size <= 0)
```

File: Objective-C/CBLReplicator.mm (hunk 5 above)
Comment (jayahariv) [reply Objective-C/CBLReplicator.mm]: will create a separate PR with replicator change.

**Pull request merged**